### PR TITLE
Add search OU import, export and search

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/8694ba058

### :video_camera: Screenshots/Screen capture

https://github.com/EyeSeeTea/user-extended-app/assets/37223065/fd686bd0-a6db-4786-bd87-c487853343bf

Screenshot 2024-05-10 at 11 02 17
Screenshot 2024-05-10 at 11 04 06

### :fire: Testing

#8694ba058

### CODE REVIEW

**Comment by ifoche** [comment]
Task linked: [CU-8694ba058 Add search, import and export "Search OU"](https://app.clickup.com/t/8694ba058)

File: src/legacy/models/userList.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ const queryFields = [
     "userGroups[id,displayName,publicAccess]",
     "organisationUnits[id,code,shortName,displayName]",
     "dataViewOrganisationUnits[id,code,shortName,displayName]",
+    "searchOrganisationUnits[id,code,shortName,displayName]",
```
Comment (eperedo) [inline src/legacy/models/userList.js:21]: Same, `teiSearchOrganisationUnits`

**Review by eperedo — CHANGES_REQUESTED** [review]
Thanks for the code @deeonwuli. I'm getting this error after selecting a org unit in the search filter: `"Unknown path property: searchOrganisationUnits.id"`

image

More details in one of my comments, let me know if you any questions.

File: src/legacy/models/userHelpers.js [hunk 2]
```diff
@@ -165,6 +165,7 @@ function getPlainUser(user, { orgUnitsField }, toArray) {
         userGroups: namesFromCollection(user.userGroups, "displayName", toArray),
         organisationUnits: namesFromCollection(user.organisationUnits, orgUnitsField, toArray),
         dataViewOrganisationUnits: namesFromCollection(user.dataViewOrganisationUnits, orgUnitsField, toArray),
+        searchOrganisationsUnits: namesFromCollection(user.dataViewOrganisationUnits, orgUnitsField, toArray),
```
[the remark this answers is not in the document]
Comment (deeonwuli) [reply src/legacy/models/userHelpers.js]: right

File: src/legacy/List/list.store.js [hunk 3]
```diff
@@ -13,7 +13,7 @@ export const columns = [
     { name: "userRoles", sortable: false },
     { name: "userGroups", sortable: false },
     { name: "organisationUnits", sortable: false },
-    { name: "dataViewOrganisationUnits", sortable: false },
```
Comment (eperedo) [inline src/legacy/List/list.store.js:16]: is `dataViewOrganisationUnits` not necessary anymore?

File: src/legacy/models/userHelpers.js [hunk 4]
```diff
@@ -165,6 +166,7 @@ function getPlainUser(user, { orgUnitsField }, toArray) {
         userGroups: namesFromCollection(user.userGroups, "displayName", toArray),
         organisationUnits: namesFromCollection(user.organisationUnits, orgUnitsField, toArray),
         dataViewOrganisationUnits: namesFromCollection(user.dataViewOrganisationUnits, orgUnitsField, toArray),
+        teiSearchOrganisationsUnits: namesFromCollection(user.dataViewOrganisationUnits, orgUnitsField, toArray),
```
Comment (eperedo) [inline src/legacy/models/userHelpers.js:169]: We can keep the entity name at the object level, but the dhis2 property name to get the value:

```tsx
searchOrganisationsUnits: namesFromCollection(user.teiSearchOrganisationUnits, orgUnitsField, toArray),
```

**Review by eperedo — CHANGES_REQUESTED** [review]
Sorry @deeonwuli but still getting the same error: `"Unknown path property: searchOrganisationUnits.id"` Check my comment for the file `Filters.component.js`

Also I cannot modify the search org. units in the import modal:

[User-Extended-App_import_csv.webm](https://github.com/EyeSeeTea/user-extended-app/assets/461124/4a0ab18e-8e14-4330-8f52-a193fb00e970)

**Review by eperedo — CHANGES_REQUESTED** [review]
We're very close @deeonwuli, right now if you try to import new users it'll send the `searchOrganisationUnits` array, but that's not right since DHIS needs the attribute `teiSearchOrganisationUnits`.

I saw a function `getUserPayloadFromPlainAttributes` in `userHelper.js` where there's a mapping before sending all the payload. I think you can add the attribute in there (if you think there's a better alternative that's ok feel free to do it the best way)

```js
function getUserPayloadFromPlainAttributes(baseUser, userFields) {
  // ...
  return {
    ...userRoot,
teiSearchOrganisationUnits: userRoot.searchOrganisationsUnits ? userRoot.searchOrganisationsUnits : undefined,
    // ...
  };
}
```

**Review by eperedo — APPROVED** [review]
Looks good, thanks @deeonwuli!

**Pull request merged**